aws/src/main/java/org/apache/iceberg/aws/glue/LockManagers.java

```diff
@@ -0,0 +1,241 @@
+   * This implementation should only be used for testing,
+   * or if the caller only needs locking within the same JVM during table commits.
+   */
+  static class InMemoryLockManager extends BaseLockManager {
+
+    private static final Map<String, DefaultLockContent> LOCKS = Maps.newConcurrentMap();
+    private static final Map<String, ScheduledFuture<?>> HEARTBEATS = Maps.newHashMap();
+
+    InMemoryLockManager(Map<String, String> properties) {
+      initialize(properties);
+    }
+
+    @VisibleForTesting
+    void acquireOnce(String entityId, String ownerId) {
+      DefaultLockContent content = LOCKS.get(entityId);
+      if (content != null && content.expireMs() > System.currentTimeMillis()) {
+        throw new IllegalStateException(String.format("Lock for %s currently held by %s, expiration: %s",
+            entityId, content.ownerId(), content.expireMs()));
+      }
+
+      long expiration = System.currentTimeMillis() + heartbeatTimeoutMs();
+      boolean succeed;
+      if (content == null) {
+        DefaultLockContent previous = LOCKS.putIfAbsent(
+            entityId, new DefaultLockContent(ownerId, expiration));
+        succeed = previous == null;
+      } else {
+        succeed = LOCKS.replace(entityId, content, new DefaultLockContent(ownerId, expiration));
+      }
+
+      if (succeed) {
+        // cleanup old heartbeat
+        if (HEARTBEATS.containsKey(entityId)) {
+          HEARTBEATS.remove(entityId).cancel(false);
+        }
+
+        HEARTBEATS.put(entityId, scheduler().scheduleAtFixedRate(() -> {
+          DefaultLockContent lastContent = LOCKS.get(entityId);
+          try {
+            long newExpiration = System.currentTimeMillis() + heartbeatTimeoutMs();
+            LOCKS.replace(entityId, lastContent, new DefaultLockContent(ownerId, newExpiration));
+          } catch (NullPointerException e) {
+            throw new RuntimeException("Cannot heartbeat to a deleted lock " + entityId, e);
+          }
+
+        }, 0, heartbeatIntervalMs(), TimeUnit.MILLISECONDS));
+
+      } else {
+        throw new IllegalStateException("Unable to acquire lock " + entityId);
+      }
+    }
+
+    @Override
+    public boolean acquire(String entityId, String ownerId) {
+      try {
+        Tasks.foreach(entityId)
+            .retry(Integer.MAX_VALUE - 1)
+            .onlyRetryOn(IllegalStateException.class)
+            .throwFailureWhenFinished()
+            .exponentialBackoff(acquireIntervalMs(), acquireIntervalMs(), acquireTimeoutMs(), 1)
+            .run(id -> acquireOnce(id, ownerId));
+        return true;
+      } catch (IllegalStateException e) {
+        return false;
+      }
+    }
+
+    @Override
+    public void release(String entityId, String ownerId) {
+      DefaultLockContent currentContent = LOCKS.get(entityId);
+      if (currentContent == null) {
+        throw new IllegalArgumentException("Cannot find lock for entity " + entityId);
```

Comment:
`IllegalArgumentException` isn't quite right. The argument is correct, it is the lock state that is the problem. `IllegalStateException` is a better one to throw for this and the next case.

We also need to consider what this exception would cause if thrown. The `release` method is currently called in a `finally` block so if it throws an exception, it will replace the real failure. I noted that above where if the lock can't be acquired, either this or the next check's exception will be thrown.

In addition to that problem, a failure here could cause duplicate data. If the table state is updated, then success _must_ be reported back. If not, then it is very likely that the operation will retry at some level and there is no guarantee that the operation is idempotent. For example, a writer moving data from Kafka to an Iceberg table will append data. If a checkpoint commit fails, it will retry. If the failure happened after the commit actually succeeded, then the append will be done twice causing duplicate data in the table.

Most of the time, a failure like that would cause the metadata file that was to be committed to be deleted, which ends up locking up the table. But this unlock happens after the delete would happen, so the table would appear correct but the operation would probably retry.

I think the solution to these problems is to make this log the error and return a boolean. If the lock was held and released, return true. If it was not held, then return false. That way the caller can decide how to handle it. And we should log the error messages you have here for context.